File: server/internal/protocol/codec.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"

	"github.com/vmihailenco/msgpack/v5"
)

var (
	// ErrFrameTooLarge indicates a frame exceeds MaxFrameSize.
	ErrFrameTooLarge = errors.New("frame too large")
	// ErrUnknownFrameType indicates an unrecognized frame type byte.
	ErrUnknownFrameType = errors.New("unknown frame type")
	// ErrIncompleteFrame indicates insufficient data to decode a frame.
	ErrIncompleteFrame = errors.New("incomplete frame")
)

// Codec handles encoding and decoding of protocol frames.
type Codec struct{}

// EncodeControl encodes a ControlMessage to msgpack bytes.
func (c *Codec) EncodeControl(msg *ControlMessage) ([]byte, error) {
	return msgpack.Marshal(msg)
}

// DecodeControl decodes a ControlMessage from msgpack bytes.
func (c *Codec) DecodeControl(data []byte) (*ControlMessage, error) {
	var msg ControlMessage
	if err := msgpack.Unmarshal(data, &msg); err != nil {
		return nil, fmt.Errorf("decode control: %w", err)
	}
	return &msg, nil
}

// EncodeDesktopFrame encodes a DesktopFrame to msgpack bytes.
func (c *Codec) EncodeDesktopFrame(f *DesktopFrame) ([]byte, error) {
	return msgpack.Marshal(f)
}

// DecodeDesktopFrame decodes a DesktopFrame from msgpack bytes.
func (c *Codec) DecodeDesktopFrame(data []byte) (*DesktopFrame, error) {
	var f DesktopFrame
	if err := msgpack.Unmarshal(data, &f); err != nil {
		return nil, fmt.Errorf("decode desktop frame: %w", err)
	}
	return &f, nil
}

// EncodeTerminalFrame encodes a TerminalFrame to msgpack bytes.
func (c *Codec) EncodeTerminalFrame(f *TerminalFrame) ([]byte, error) {
	return msgpack.Marshal(f)
}

// DecodeTerminalFrame decodes a TerminalFrame from msgpack bytes.
func (c *Codec) DecodeTerminalFrame(data []byte) (*TerminalFrame, error) {
	var f TerminalFrame
	if err := msgpack.Unmarshal(data, &f); err != nil {
		return nil, fmt.Errorf("decode terminal frame: %w", err)
	}
	return &f, nil
}

// EncodeFileFrame encodes a FileFrame to msgpack bytes.
func (c *Codec) EncodeFileFrame(f *FileFrame) ([]byte, error) {
	return msgpack.Marshal(f)
}

// DecodeFileFrame decodes a FileFrame from msgpack bytes.
func (c *Codec) DecodeFileFrame(data []byte) (*FileFrame, error) {
	var f FileFrame
	if err := msgpack.Unmarshal(data, &f); err != nil {
		return nil, fmt.Errorf("decode file frame: %w", err)
	}
	return &f, nil
}

// WriteFrame writes a typed frame to a writer.
// Wire format: [type_byte][4-byte big-endian length][payload]
// Ping/Pong are written as a single type byte with no length or payload.
func (c *Codec) WriteFrame(w io.Writer, frameType byte, payload []byte) error {
	if frameType == FramePing || frameType == FramePong {
		_, err := w.Write([]byte{frameType})
		return err
	}

	if len(payload) > MaxFrameSize {
		return fmt.Errorf("%w: %d bytes (max %d)", ErrFrameTooLarge, len(payload), MaxFrameSize)
	}

	header := [5]byte{frameType}
	binary.BigEndian.PutUint32(header[1:], uint32(len(payload)))
	if _, err := w.Write(header[:]); err != nil {
		return err
	}
	if len(payload) > 0 {
		_, err := w.Write(payload)
		return err
	}
	return nil
}

// ReadFrame reads a typed frame from a reader.
// Returns the frame type, payload, and any error.
func (c *Codec) ReadFrame(r io.Reader) (byte, []byte, error) {
	var typeBuf [1]byte
	if _, err := io.ReadFull(r, typeBuf[:]); err != nil {
		return 0, nil, err
	}
	frameType := typeBuf[0]

	if frameType == FramePing || frameType == FramePong {
		return frameType, nil, nil
	}

	if frameType != FrameControl && frameType != FrameDesktop &&
		frameType != FrameTerminal && frameType != FrameFile {
		return 0, nil, fmt.Errorf("%w: 0x%02x", ErrUnknownFrameType, frameType)
	}

	var lengthBuf [4]byte
	if _, err := io.ReadFull(r, lengthBuf[:]); err != nil {
		return 0, nil, fmt.Errorf("%w: %v", ErrIncompleteFrame, err)
	}
	length := binary.BigEndian.Uint32(lengthBuf[:])

	if int(length) > MaxFrameSize {
		return 0, nil, fmt.Errorf("%w: %d bytes (max %d)", ErrFrameTooLarge, length, MaxFrameSize)
	}

	payload := make([]byte, length)
	if length > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return 0, nil, fmt.Errorf("%w: %v", ErrIncompleteFrame, err)
		}
	}
	return frameType, payload, nil
}

// EncodeHandshake encodes a handshake message to binary format.
// Wire format matches the Rust HandshakeMessage::encode_binary.
func EncodeHandshake(msgType byte, payload []byte) []byte {
	buf := make([]byte, 1+len(payload))
	buf[0] = msgType
	copy(buf[1:], payload)
	return buf
}

// EncodeServerHello encodes a ServerHello handshake message.
func EncodeServerHello(nonce [32]byte, certHash [48]byte) []byte {
	buf := make([]byte, 81)
	buf[0] = MsgServerHello
	copy(buf[1:33], nonce[:])
	copy(buf[33:81], certHash[:])
	return buf
}

// EncodeAgentHello encodes an AgentHello handshake message.
func EncodeAgentHello(nonce [32]byte, agentCertHash [48]byte) []byte {
	buf := make([]byte, 81)
	buf[0] = MsgAgentHello
	copy(buf[1:33], nonce[:])
	copy(buf[33:81], agentCertHash[:])
	return buf
}

// DecodeHandshakeType returns the type byte from a handshake message.
func DecodeHandshakeType(data []byte) (byte, error) {
	if len(data) == 0 {
		return 0, fmt.Errorf("empty handshake data")
	}
	return data[0], nil
}

// DecodeServerHello decodes a ServerHello handshake message.
func DecodeServerHello(data []byte) (nonce [32]byte, certHash [48]byte, err error) {
	if len(data) != 81 {
		return nonce, certHash, fmt.Errorf("ServerHello must be 81 bytes, got %d", len(data))
	}
	if data[0] != MsgServerHello {
		return nonce, certHash, fmt.Errorf("expected ServerHello type 0x10, got 0x%02x", data[0])
	}
	copy(nonce[:], data[1:33])
	copy(certHash[:], data[33:81])
	return nonce, certHash, nil
}

File: web/src/lib/protocol/codec.ts
import { encode, decode } from '@msgpack/msgpack';
import type { Frame, ControlMessage, DesktopFrame, TerminalFrame, FileFrame } from './types';
import {
  FRAME_CONTROL,
  FRAME_DESKTOP,
  FRAME_TERMINAL,
  FRAME_FILE,
  FRAME_PING,
  FRAME_PONG,
  MAX_FRAME_SIZE,
} from './types';

/** Encode a Frame to wire format: [1-byte type][4-byte BE length][msgpack payload]. */
export function encodeFrame(frame: Frame): Uint8Array<ArrayBuffer> {
  if (frame.type === FRAME_PING) return new Uint8Array([FRAME_PING]);
  if (frame.type === FRAME_PONG) return new Uint8Array([FRAME_PONG]);

  let payload: Uint8Array;
  switch (frame.type) {
    case FRAME_CONTROL:
      payload = encode(frame.message);
      break;
    case FRAME_DESKTOP:
      payload = encode(frame.frame);
      break;
    case FRAME_TERMINAL:
      payload = encode(frame.frame);
      break;
    case FRAME_FILE:
      payload = encode(frame.frame);
      break;
    default:
      throw new Error(`unexpected frame type: ${(frame as { type: number }).type}`);
  }

  if (payload.length > MAX_FRAME_SIZE) {
    throw new Error(`frame payload too large: ${payload.length} > ${MAX_FRAME_SIZE}`);
  }

  const buf = new Uint8Array(5 + payload.length);
  buf[0] = frame.type;
  const view = new DataView(buf.buffer);
  view.setUint32(1, payload.length, false); // big-endian
  buf.set(payload, 5);
  return buf;
}

/** Decode a Frame from wire format. Returns the frame and number of bytes consumed. */
export function decodeFrame(data: Uint8Array): { frame: Frame; bytesConsumed: number } {
  if (data.length === 0) {
    throw new Error('incomplete frame: empty data');
  }

  const typeByte = data[0] as number;

  if (typeByte === FRAME_PING) return { frame: { type: FRAME_PING }, bytesConsumed: 1 };
  if (typeByte === FRAME_PONG) return { frame: { type: FRAME_PONG }, bytesConsumed: 1 };

  if (
    typeByte !== FRAME_CONTROL &&
    typeByte !== FRAME_DESKTOP &&
    typeByte !== FRAME_TERMINAL &&
    typeByte !== FRAME_FILE
  ) {
    throw new Error(`unknown frame type: 0x${typeByte.toString(16).padStart(2, '0')}`);
  }

  if (data.length < 5) {
    throw new Error(`incomplete frame: need 5 bytes, have ${data.length}`);
  }

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const length = view.getUint32(1, false); // big-endian

  if (length > MAX_FRAME_SIZE) {
    throw new Error(`frame too large: ${length} > ${MAX_FRAME_SIZE}`);
  }

  const total = 5 + length;
  if (data.length < total) {
    throw new Error(`incomplete frame: need ${total} bytes, have ${data.length}`);
  }

  const payload = data.subarray(5, total);
  const decoded = decode(payload);

  let frame: Frame;
  switch (typeByte) {
    case FRAME_CONTROL:
      frame = { type: FRAME_CONTROL, message: decoded as ControlMessage };
      break;
    case FRAME_DESKTOP:
      frame = { type: FRAME_DESKTOP, frame: decoded as DesktopFrame };
      break;
    case FRAME_TERMINAL:
      frame = { type: FRAME_TERMINAL, frame: decoded as TerminalFrame };
      break;
    case FRAME_FILE:
      frame = { type: FRAME_FILE, frame: decoded as FileFrame };
      break;
    default:
      throw new Error(`unexpected frame type: 0x${(typeByte as number).toString(16).padStart(2, '0')}`);
  }

  return { frame, bytesConsumed: total };
}
